### PR TITLE
torch_xla trainer uses sharding config

### LINKED ISSUE #21: Support changing the sharding from a model config
Once it's possible to configure the shapes/rope embeddings/norms of a model, we should also support configuring the sharding of various parameters from a config file. The config file could be in Python or in `.yaml` etc.

One idea for reference: https://github.com/AI-Hypercomputer/jetstream-pytorch/blob/main/default_shardings/llama.yaml

The expectation is that someone won't specify `--fsdp` or `--spmd_2d_sharding` on the CLI. Instead, they can specify a sharding config name/path. We'll expect to implement FSDP sharding via the config. And we'll be able to test a variety of sharding configs.

### PR DESCRIPTION
**Pull request by tengyifei**
This PR replaces FSDPv2 and hardcoded sharding specs with sharding config files.

Tests:

- Added a test that compares FSDPv2 and config based sharder numerics and sharding specs for a 2 layer Llama and 2 layer Mixtral.
- I compared the loss curve and running time for Llama on v6e-8, and Mixtral on v6e-8.
- I also compared them on a Trillium pod on xpk.

## Llama

Llama FSDPv2:

```
 - Step: 0, loss: 12.5718, trace time: 134588.13 ms
 - Step: 10, loss: 10.9212, trace time: 878.29 ms
 - Step: 20, loss: 9.1416, trace time: 880.09 ms
 - Step: 30, loss: 8.2824, trace time: 878.94 ms
 - Step: 40, loss: 7.0730, trace time: 883.22 ms
 - Step: 50, loss: 6.8937, trace time: 862.05 ms
 - Step: 60, loss: 6.9061, trace time: 857.86 ms
 - Step: 70, loss: 6.8793, trace time: 874.83 ms
 - Step: 80, loss: 6.8040, trace time: 145796.24 ms
 - Step: 90, loss: 6.7863, trace time: 913.06 ms
 - Finished training run
 - Step duration: 3.695 s
```

Llama Config based:

```
 - Step: 0, loss: 12.5718, trace time: 119801.30 ms
 - Step: 10, loss: 10.9285, trace time: 982.66 ms
 - Step: 20, loss: 9.2224, trace time: 961.95 ms
 - Step: 30, loss: 7.4156, trace time: 979.38 ms
 - Step: 40, loss: 6.9384, trace time: 974.15 ms
 - Step: 50, loss: 6.8775, trace time: 952.50 ms
 - Step: 60, loss: 6.8922, trace time: 963.18 ms
 - Step: 70, loss: 6.8683, trace time: 938.67 ms
 - Step: 80, loss: 6.7947, trace time: 115190.57 ms
 - Step: 90, loss: 6.7745, trace time: 984.25 ms
 - Finished training run
 - Step duration: 3.703 s
```

The delta in loss and step time is ~0.2%. The are not identical despite the identical sharding. Therefore, I also trained this on a v4-8 and the loss numbers also changes quite a bit, so this should be within acceptible noise interval.

For reference, FSDPv2 on v4-8 curve looks like this:

```
 - Step: 0, loss: 12.5720, trace time: 123530.85 ms
 - Step: 10, loss: 11.5086, trace time: 1688.35 ms
 - Step: 20, loss: 9.5840, trace time: 1726.88 ms
 - Step: 30, loss: 8.7745, trace time: 1757.86 ms
 - Step: 40, loss: 7.1194, trace time: 1740.76 ms
 - Step: 50, loss: 6.9021, trace time: 1717.90 ms
 - Step: 60, loss: 6.9035, trace time: 2015.98 ms
 - Step: 70, loss: 6.8790, trace time: 1973.11 ms
 - Step: 80, loss: 6.8034, trace time: 130805.92 ms
 - Step: 90, loss: 6.7861, trace time: 1738.55 ms
```

This suggests to me that the loss delta is acceptable.

I also ran Llama 3-8B on xpk and got the same step time as https://github.com/AI-Hypercomputer/torchprime/issues/16#issuecomment-2638386904.

## Mixtral

Mixtral 8x7B with FSDPv2:

```
 - Step: 0, loss: 11.1757, trace time: 440648.42 ms
 - Step: 10, loss: 10.4170, trace time: 9153.15 ms
 - Step: 20, loss: 8.5687, trace time: 9129.96 ms
 - Step: 30, loss: 7.6070, trace time: 9858.94 ms
 - Step: 40, loss: 7.1508, trace time: 9779.03 ms
 - Step: 50, loss: 7.0312, trace time: 9774.99 ms
 - Step: 60, loss: 6.9317, trace time: 9829.54 ms
 - Step: 70, loss: 6.7952, trace time: 9976.81 ms
 - Step: 80, loss: 6.6540, trace time: 450864.63 ms
 - Step: 90, loss: 6.6430, trace time: 9370.79 ms
 - Finished training run
 - Step duration: 10.001 s
```

Mixtral 8x7B with config sharding:

```
 - Step: 0, loss: 11.1757, trace time: 413730.83 ms
 - Step: 10, loss: 10.4113, trace time: 3784.12 ms
 - Step: 20, loss: 8.0267, trace time: 3779.08 ms
 - Step: 30, loss: 7.2595, trace time: 3821.59 ms
 - Step: 40, loss: 7.1256, trace time: 3824.65 ms
 - Step: 50, loss: 7.2663, trace time: 3515.31 ms
 - Step: 60, loss: 6.8813, trace time: 3360.15 ms
 - Step: 70, loss: 6.7367, trace time: 3341.72 ms
 - Step: 80, loss: 6.6374, trace time: 432356.79 ms
 - Step: 90, loss: 6.6372, trace time: 3431.99 ms
 - Finished training run
 - Step duration: 9.951 s
```

The delta is similar to Llama.

I also ran Mixtral 8x7B on xpk and got similar step time as a run from the main branch (using FSDPv2) using pdbs=2:

- FSDPv2: 8.144 s
- Config based: 8.153 s
- Delta: ~0.1%

Fixes #21.

### CODE REVIEW

**Review by bhavya01 — APPROVED** [review]
Awesome! Thanks for the diligent testing.
